From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: David Mayr <davidliebtkekse@gmail.com>
Date: Sat, 16 Dec 2023 10:40:29 +0100
Subject: [PATCH] add number format api

== AT ==
public net.minecraft.network.chat.numbers.FixedFormat value
public net.minecraft.network.chat.numbers.StyledFormat style

diff --git a/src/main/java/io/papermc/paper/util/PaperScoreboardFormat.java b/src/main/java/io/papermc/paper/util/PaperScoreboardFormat.java
new file mode 100644
index 0000000000000000000000000000000000000000..0aff8db3eec522bc1ad9c77a559b3dc0be3b89c7
--- /dev/null
+++ b/src/main/java/io/papermc/paper/util/PaperScoreboardFormat.java
@@ -0,0 +1,52 @@
+package io.papermc.paper.util;
+
+
+import io.papermc.paper.adventure.AdventureCodecs;
+import io.papermc.paper.scoreboard.numbers.NumberFormat;
+import net.minecraft.Util;
+import net.minecraft.nbt.NbtOps;
+import net.minecraft.nbt.Tag;
+
+public class PaperScoreboardFormat {
+
+    private PaperScoreboardFormat() {}
+
+
+    public static net.minecraft.network.chat.numbers.NumberFormat asVanilla(NumberFormat format) {
+        net.minecraft.network.chat.numbers.NumberFormat vanilla;
+        if (format instanceof io.papermc.paper.scoreboard.numbers.StyledFormat styled) {
+            Tag encoded = Util.getOrThrow(AdventureCodecs.STYLE_MAP_CODEC.codec().encodeStart(NbtOps.INSTANCE, styled.style()), IllegalStateException::new);
+
+            net.minecraft.network.chat.Style vanillaStyle = Util.getOrThrow(net.minecraft.network.chat.Style.Serializer.CODEC
+                .parse(NbtOps.INSTANCE, encoded), IllegalStateException::new);
+
+            vanilla = new net.minecraft.network.chat.numbers.StyledFormat(vanillaStyle);
+        } else if (format instanceof io.papermc.paper.scoreboard.numbers.FixedFormat fixed) {
+            vanilla = new net.minecraft.network.chat.numbers.FixedFormat(io.papermc.paper.adventure.PaperAdventure
+                .asVanilla(fixed.component()));
+        } else if(format instanceof io.papermc.paper.scoreboard.numbers.BlankFormat) {
+            vanilla = net.minecraft.network.chat.numbers.BlankFormat.INSTANCE;
+        } else {
+            throw new IllegalArgumentException("Unknown format type " + format.getClass());
+        }
+
+        return vanilla;
+    }
+
+    public static NumberFormat asPaper(net.minecraft.network.chat.numbers.NumberFormat vanilla) {
+        if (vanilla instanceof net.minecraft.network.chat.numbers.StyledFormat styled) {
+            Tag encoded = Util.getOrThrow(net.minecraft.network.chat.Style.Serializer.CODEC.encodeStart(NbtOps.INSTANCE, styled.style), IllegalStateException::new);
+
+            net.kyori.adventure.text.format.Style adventureStyle = Util.getOrThrow(AdventureCodecs.STYLE_MAP_CODEC.codec()
+                .parse(NbtOps.INSTANCE, encoded), IllegalStateException::new);
+
+            return new io.papermc.paper.scoreboard.numbers.StyledFormat(adventureStyle);
+        } else if (vanilla instanceof net.minecraft.network.chat.numbers.FixedFormat fixed) {
+            return new io.papermc.paper.scoreboard.numbers.FixedFormat(io.papermc.paper.adventure.PaperAdventure.asAdventure(fixed.value));
+        } else if(vanilla instanceof net.minecraft.network.chat.numbers.BlankFormat) {
+            return io.papermc.paper.scoreboard.numbers.BlankFormat.blank();
+        }
+
+        throw new IllegalArgumentException("Unknown format type " + vanilla.getClass());
+    }
+}
diff --git a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftObjective.java b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftObjective.java
index 3157f3d2f9ce7af4a763203672817a7f5c7bd4fb..bdd2ed0aa62eea87d356d2229bb7fe41ba293e1d 100644
--- a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftObjective.java
+++ b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftObjective.java
@@ -152,6 +152,30 @@ final class CraftObjective extends CraftScoreboardComponent implements Objective
     }
     // Paper end
 
+    //Paper start - add number format
+    @Override
+    public io.papermc.paper.scoreboard.numbers.NumberFormat getNumberFormat() {
+        net.minecraft.network.chat.numbers.NumberFormat vanilla = this.objective.numberFormat();
+        
+        if(vanilla == null) {
+            return null;
+        }
+
+        return io.papermc.paper.util.PaperScoreboardFormat.asPaper(vanilla);
+    }
+
+
+    @Override
+    public void setNumberFormat(io.papermc.paper.scoreboard.numbers.NumberFormat format) {
+        if(format == null) {
+            this.objective.setNumberFormat(null);
+            return;
+        }
+
+        this.objective.setNumberFormat(io.papermc.paper.util.PaperScoreboardFormat.asVanilla(format));
+    }
+    //Paper end - add number format
+
     @Override
     public void unregister() {
         CraftScoreboard scoreboard = this.checkState();
diff --git a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScore.java b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScore.java
index be020d6b686c1ad5bd8b7cee0b6050304d434022..e07d8278722123341c9a60e99dccacab37a6c21e 100644
--- a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScore.java
+++ b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScore.java
@@ -57,6 +57,35 @@ final class CraftScore implements Score {
         this.objective.checkState().board.getOrCreatePlayerScore(this.entry, this.objective.getHandle()).set(score);
     }
 
+
+    //Paper start - add number format
+    @Override
+    public io.papermc.paper.scoreboard.numbers.NumberFormat getNumberFormat() {
+        net.minecraft.network.chat.numbers.NumberFormat vanilla = this.objective.checkState().board
+            .getPlayerScoreInfo(this.entry, this.objective.getHandle()).numberFormat();
+
+        if(vanilla == null) {
+            return null;
+        }
+
+        return io.papermc.paper.util.PaperScoreboardFormat.asPaper(vanilla);
+    }
+
+
+    @Override
+    public void setNumberFormat(io.papermc.paper.scoreboard.numbers.NumberFormat format) {
+        final net.minecraft.world.scores.ScoreAccess access = this.objective.checkState()
+            .board.getOrCreatePlayerScore(this.entry, this.objective.getHandle());
+
+        if(format == null) {
+            access.numberFormatOverride(null);
+            return;
+        }
+
+        access.numberFormatOverride(io.papermc.paper.util.PaperScoreboardFormat.asVanilla(format));
+    }
+    //Paper end - add number format
+
     @Override
     public boolean isScoreSet() {
         Scoreboard board = this.objective.checkState().board;
diff --git a/src/test/java/io/papermc/paper/scoreboard/NumberFormatTest.java b/src/test/java/io/papermc/paper/scoreboard/NumberFormatTest.java
new file mode 100644
index 0000000000000000000000000000000000000000..7651c8d5af82b250888867702512e0727c54e93c
--- /dev/null
+++ b/src/test/java/io/papermc/paper/scoreboard/NumberFormatTest.java
@@ -0,0 +1,24 @@
+package io.papermc.paper.scoreboard;
+
+import io.papermc.paper.util.PaperScoreboardFormat;
+import net.minecraft.network.chat.numbers.NumberFormat;
+import net.minecraft.network.chat.numbers.StyledFormat;
+import org.bukkit.support.AbstractTestingBase;
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class NumberFormatTest extends AbstractTestingBase {
+
+    @Test
+    public void testStyledFormatDefaults() {
+        StyledFormat noneConverted = (StyledFormat) PaperScoreboardFormat.asVanilla(io.papermc.paper.scoreboard.numbers.StyledFormat.NO_STYLE);
+        assertEquals(noneConverted.style, StyledFormat.NO_STYLE.style, "The paper scoreboard number format \"NO_STYLE\" does not match vanilla");
+
+        StyledFormat sideBarConverted = (StyledFormat) PaperScoreboardFormat.asVanilla(io.papermc.paper.scoreboard.numbers.StyledFormat.SIDEBAR_DEFAULT);
+        assertEquals(sideBarConverted.style, StyledFormat.SIDEBAR_DEFAULT.style, "The paper scoreboard number format \"SIDEBAR_DEFAULT\" does not match vanilla");
+
+        StyledFormat playerListConverted = (StyledFormat) PaperScoreboardFormat.asVanilla(io.papermc.paper.scoreboard.numbers.StyledFormat.PLAYER_LIST_DEFAULT);
+        assertEquals(playerListConverted.style, StyledFormat.PLAYER_LIST_DEFAULT.style, "The paper scoreboard number format \"PLAYER_LIST_DEFAULT\" does not match vanilla");
+    }
+
+}
